### ISSUE TITLE
aws cant create RDS db.t2.medium multi az zone

### ISSUE THREAD

**Issue by sukritpol** [hashicorp/terraform#11407]
Could you help us , we cant create RDS db.t2.medium multi az zone  but can change in aws console
~~~
Error creating DB Instance: InvalidParameterCombination: Requesting a specific availability zone is not valid for Multi-AZ instances.
~~~

**Comment by stack72** [comment]
Hi @sukritpol 

In order to help us diagnose your issue, please can you post your terraform config (minus any secrets)

Thanks

Paul

**Comment by sukritpol** [comment]
**terraform.tfvars** 
~~~
rds_db_port = "3306"
rds_instance_type = "db.t2.medium"
rds_master_password = "****"
rds_master_username = "root"
rds_multi_az = "false" >> can not use true
rds_service_vault = "xxxx"
rds_dbname_vault = "xxx"
rds_iops = "2000"
rds_storage_type = "io1"
~~~

 **database.tf**
~~~
#--------------------------------------------------------------
# This module creates all resources necessary for a RDS instance
#--------------------------------------------------------------

variable "allocated_storage" {default = 200}
variable "abbreviation" {}
variable "azs" {}
variable "db_name" {}
variable "db_instance_type" {}
variable "db_master_username" {}
variable "db_master_password" {}
variable "db_multi_az" {}
variable "db_port" {}
variable "iops" {default = 2000}
variable "jump_box_private_ip" {}
variable "name" {}
variable "private_subnet_cidrs" {}
variable "private_subnet_ids" {}
variable "tag_bu" {}
variable "tag_country" {}
variable "tag_environment" {}
variable "tag_product" {}
variable "vpc_id" {}
variable "storage_type" {default = "gp2"}


variable "rds-gp2" {
	default {
		alp = "1"
		stg = "1"
		prd = "0"
	}
}

variable "rds-iops" {
	default {
		alp = "0"
		stg = "0"
		prd = "1"
	}
}
#
# create rds security group
#
resource "aws_security_group" "rds" {
  name = "${var.name}-rds-sg"
  description = "RDS Security Group"
  vpc_id = "${var.vpc_id}"

  lifecycle { create_before_destroy = true }
  ingress {
    from_port = "${var.db_port}"
    to_port = "${var.db_port}"
    protocol = "tcp"
    cidr_blocks = ["${split(",", var.private_subnet_cidrs)}"]
  }

  ingress {
    from_port = "${var.db_port}"
    to_port = "${var.db_port}"
    protocol = "tcp"
    cidr_blocks = ["${var.jump_box_private_ip}/32"]
  }

  egress {
      from_port = 0
      to_port = 0
      protocol = "-1"
      cidr_blocks = ["0.0.0.0/0"]
  }

  tags {
    AWSComponent = "sg"
    BU = "${var.tag_bu}"
    Country = "${var.tag_country}"
    Environment = "${var.tag_environment}"
    Name = "${var.name}-rds-sg"
    NetworkSegment = "${element(split(",", var.azs), 0)}"
    Product = "${var.tag_product}"
    Service = "db"
    ServiceComponent = ""
  }
}

#--------------------------------------------------------------
# RDS Subnet Group
#--------------------------------------------------------------
resource "aws_db_subnet_group" "rds" {
  name = "${var.name}-db-sng"
  description = "RDS Subnet Group"
  subnet_ids = ["${split(",", var.private_subnet_ids)}"]

  tags {
    AWSComponent = "sng"
    BU = "${var.tag_bu}"
    Country = "${var.tag_country}"
    Environment = "${var.tag_environment}"
    Name = "${var.name}-db"
    NetworkSegment = ""
    Product = "${var.tag_product}"
    Service = "db"
    ServiceComponent = ""
    }
}

resource "aws_db_parameter_group" "rds" {
  name = "${var.name}-parametergroup-db-myl"
  family = "mysql5.6"

  parameter {
    name = "local_infile"
    value = "0"
  }

  parameter {
    name = "log_warnings"
    value = "2"
  }
  parameter {
    name = "read_only"
    value = "{TrueIfReplica}"
  }
  parameter {
    name = "sync_binlog"
    value = "1"
  }
  parameter {
    name = "time_zone"
    value = "UTC"
  }
  parameter {
    name = "slow_query_log"
    value = "1"
  }

  //need to restart DB
  parameter {
    name = "character-set-client-handshake"
    value = "1"
    apply_method = "pending-reboot"
  }
  parameter {
    name = "character_set_client"
    value = "utf8"
    apply_method = "pending-reboot"
  }
  parameter {
    name = "character_set_connection"
    value = "utf8"
    apply_method = "pending-reboot"
  }
  parameter {
    name = "character_set_database"
    value = "utf8"
    apply_method = "pending-reboot"
  }
  parameter {
    name = "character_set_filesystem"
    value = "binary"
    apply_method = "pending-reboot"
  }
  parameter {
    name = "character_set_results"
    value = "utf8"
    apply_method = "pending-reboot"
  }
  parameter {
    name = "character_set_server"
    value = "utf8"
    apply_method = "pending-reboot"
  }
  parameter {
    name = "collation_connection"
    value = "utf8_unicode_ci"
    apply_method = "pending-reboot"
  }
//  parameter {
//    name = "skip-slave-start"
//    value = "1"
//  }
}

resource "aws_db_option_group" "rds" {
  name = "${var.name}-option-db-myl"
  option_group_description = "${var.name}-option-db-myl"
  engine_name = "mysql"
  major_engine_version = "5.6"

  option {
    option_name = "MARIADB_AUDIT_PLUGIN"
    option_settings {
      name = "SERVER_AUDIT_FILE_ROTATIONS"
      value = "50"
    }
    option_settings {
      name = "SERVER_AUDIT_FILE_ROTATE_SIZE"
      value = "5242880"
    }
    option_settings {
      name = "SERVER_AUDIT_INCL_USERS"
      value = "-"
    }
    option_settings {
      name = "SERVER_AUDIT_EXCL_USERS"
      value = "-"
    }
  }
}

#
# Create rds instance
#
resource "aws_db_instance" "rds" {
  count = "${lookup(var.rds-gp2, var.tag_environment)}"
  allocated_storage = "${var.allocated_storage}"
  parameter_group_name = "${aws_db_parameter_group.rds.name}"
  engine = "mysql"
  engine_version = "5.6.29"
  identifier = "${var.name}-db" # no longer than 15 letters
  instance_class = "${var.db_instance_type}"
  storage_type = "${var.storage_type}"
//  iops = "${var.iops}"
  username = "${var.db_master_username}"
  password = "${var.db_master_password}"
  backup_retention_period = "7"
  backup_window = "17:00-17:30"
  port = "${var.db_port}"
  maintenance_window = "sat:19:00-sat:19:30"
  option_group_name = "${aws_db_option_group.rds.name}"
  availability_zone  = "${element(split(",", var.azs), 0)}"
  name = "${var.db_name}"

  # Multi-AZ is not currently available in ap-southeast-1 (8-Jun-16)
  multi_az = "${var.db_multi_az}"

  vpc_security_group_ids = ["${aws_security_group.rds.id}"]
  db_subnet_group_name = "${aws_db_subnet_group.rds.name}"

  apply_immediately = "true"
//  apply_immediately = "false"
  auto_minor_version_upgrade = "true"
  allow_major_version_upgrade = "false"
  tags {
    AWSComponent = "rds"
    BU = "${var.tag_bu}"
    Country = "${var.tag_country}"
    Environment = "${var.tag_environment}"
    Name = "${var.name}-db"
    NetworkSegment = "${element(split(",", var.azs), 0)}"
    Product = "${var.tag_product}"
    Service = "db"
    ServiceComponent = ""
  }
}

resource "aws_db_instance" "rds-iops" {
  count = "${lookup(var.rds-iops, var.tag_environment)}"
  allocated_storage = "${var.allocated_storage}"
  parameter_group_name = "${aws_db_parameter_group.rds.name}"
  engine = "mysql"
  engine_version = "5.6.29"
  identifier = "${var.name}-db" # no longer than 15 letters
  instance_class = "${var.db_instance_type}"
  storage_type = "${var.storage_type}"
  iops = "${var.iops}"
  username = "${var.db_master_username}"
  password = "${var.db_master_password}"
  backup_retention_period = "7"
  backup_window = "17:00-17:30"
  port = "${var.db_port}"
  maintenance_window = "sat:19:00-sat:19:30"
  option_group_name = "${aws_db_option_group.rds.name}"
  availability_zone  = "${element(split(",", var.azs), 0)}"
  name = "${var.db_name}"

  # Multi-AZ is not currently available in ap-southeast-1 (8-Jun-16)
  multi_az = "${var.db_multi_az}"

  vpc_security_group_ids = ["${aws_security_group.rds.id}"]
  db_subnet_group_name = "${aws_db_subnet_group.rds.name}"

  apply_immediately = "true"
//  apply_immediately = "false"
  auto_minor_version_upgrade = "true"
  allow_major_version_upgrade = "false"
  tags {
    AWSComponent = "rds"
    BU = "${var.tag_bu}"
    Country = "${var.tag_country}"
    Environment = "${var.tag_environment}"
    Name = "${var.name}-db"
    NetworkSegment = "${element(split(",", var.azs), 0)}"
    Product = "${var.tag_product}"
    Service = "db"
    ServiceComponent = ""
  }
}

~~~

**Comment by josh-padnick** [comment]
We ([Gruntwork](http://gruntwork.io)) noticed that we were getting a similar issue with a slightly different error message:

```
aws_db_instance.primary: Error creating DB Instance: InvalidParameterCombination: To enable encryption at rest StorageEncrypted must be set to 'true'
```

This occurred when the `storage_encrypted` property of an `aws_db_instance` resource was set to `false`. I noticed that we also set the `kms_key_id` property to the empty string. 

All of this is consistent with our general observation that over the past few months, AWS seems to be erroring out when you submit a request with properties that it ignores but which have non-empty values. I think the solution here is for Terraform to have a first-class distinction between `unspecified` and empty values. Properties that are set to `unspecified` simply wouldn't be submitted to AWS.
